google-cloud-spanner/lib/google/cloud/spanner/convert.rb

```diff
@@ -140,42 +77,110 @@ def raw_to_value obj
                 obj.to_time.utc.strftime("%FT%T.%NZ"))
-            elsif Date === obj
+            when Date
               Google::Protobuf::Value.new string_value: obj.to_s
-            elsif Array === obj
+            when Array
+              arr_field = nil
+              arr_field = field.first if Array === field
               Google::Protobuf::Value.new list_value:
                 Google::Protobuf::ListValue.new(values:
-                  obj.map { |o| raw_to_value(o) })
-            elsif Hash === obj
-              Google::Protobuf::Value.new struct_value:
-                Google::Protobuf::Struct.new(fields:
-                  Hash[obj.map { |k, v| [String(k), raw_to_value(v)] }])
-            elsif obj.respond_to?(:read) && obj.respond_to?(:rewind)
-              obj.rewind
-              content = obj.read.force_encoding("ASCII-8BIT")
-              encoded_content = Base64.strict_encode64(content)
-              Google::Protobuf::Value.new(string_value: encoded_content)
+                  obj.map { |o| object_to_grpc_value(o, arr_field) })
+            when Hash
+              if field.is_a? Fields
+                field.struct(obj).to_grpc_value
+              else
+                raise ArgumentError,
+                      "A hash value cannot be set to type #{field}."
+              end
             else
-              raise ArgumentError,
-                    "A value of type #{obj.class} is not supported."
+              if obj.respond_to?(:read) && obj.respond_to?(:rewind)
+                obj.rewind
+                content = obj.read.force_encoding("ASCII-8BIT")
+                encoded_content = Base64.strict_encode64(content)
+                Google::Protobuf::Value.new(string_value: encoded_content)
+              else
+                raise ArgumentError,
+                      "A value of type #{obj.class} is not supported."
+              end
             end
           end
 
-          def row_to_pairs row_types, row
-            row_types.zip(row).map do |field, value|
-              [field.name.to_sym, value_to_raw(value, field.type)]
+          def field_for_object obj
+            case obj
+            when NilClass
+              raise ArgumentError, "Cannot determine type for nil values."
+            when String
+              :STRING
+            when Symbol
+              :STRING
+            when TrueClass
+              :BOOL
+            when FalseClass
+              :BOOL
+            when Integer
+              :INT64
+            when Numeric
+              :FLOAT64
+            when Time
+              :TIMESTAMP
+            when DateTime
+              :TIMESTAMP
+            when Date
+              :DATE
+            when Array
+              if obj.empty?
+                raise ArgumentError,
+                      "Cannot determine type for empty array values."
+              end
+              non_nil_fields = obj.compact.map { |e| field_for_object e }.compact
```

Comment:
For input:

 ```
client  = spanner.client instance_id, database_id
name_type = client.fields(FirstName: :STRING, LastName: :STRING)
band_members = [name_type.struct(["Elena", "Campbell"]), name_type.struct(["Gabriel", "Wright"])];

client.execute("SELECT SingerId FROM Singers WHERE " \
                  "STRUCT<FirstName STRING, LastName STRING>(FirstName, LastName) IN UNNEST(@names)",
                 params: { names: band_members }).rows.each do |row|
    puts "#{row[:SingerId]}"
                 end
```

This compact does not seem to be working correctly and returns:
```
((FirstName):STRING, (LastName):STRING)
((FirstName):STRING, (LastName):STRING)
```
And hence throws the error : "Cannot determine type for array of different values"

Can you add a test case that uses an array of struct ?

@jsimonweb FYI

Reply:
So after discussing this with @quartzmo we decided that it would be acceptable to use Hash-key comparison with `Fields` since it is an immutable object. This made the code change to fix this very simple. I've added code coverage with this change as well to make sure the `Cannot determine type for array of different values` error was not being raised.